src/main/scala/org/constellation/p2p/Download.scala

```diff
@@ -14,69 +18,183 @@ import scala.concurrent.duration._
+    } yield ()
+
+  private def initDownloadingProcess: IO[Unit] =
+    IO(logger.info("Download started"))
+      .flatMap(_ => setNodeState(NodeState.DownloadInProgress))
+      .flatMap(_ => requestForFaucet)
+      .map(_ => ())
+
+  private def downloadAndAcceptGenesis: IO[GenesisObservation] =
+    (dao.peerManager ? APIBroadcast(_.getNonBlocking[Option[GenesisObservation]]("genesis")))
+      .mapTo[Map[Schema.Id, Future[Option[GenesisObservation]]]]
+      .flatMap(m => Future.find(m.values.toList)(_.nonEmpty).map(_.flatten.get))
+      .toIO
+      .flatMap(updateMetricAndPass("downloadedGenesis", "true"))
+      .flatMap(genesis => IO(dao.acceptGenesis(genesis)).map(_ => genesis))
+
+  private def waitForPeers(): IO[Unit] =
+    IO(logger.info(s"Waiting ${waitForPeersDelay.toString()} for peers"))
+      .flatMap(_ => IO.sleep(waitForPeersDelay))
+
+  private def getReadyPeers(): IO[Peers] =
+    (dao.peerManager ? GetPeerInfo)
+      .mapTo[Map[Schema.Id, PeerData]]
+      .toIO
+      .map(_.filter(_._2.peerMetadata.nodeState == NodeState.Ready))
+
+  private def getSnapshotClient(peers: Peers) = IO(peers.head._2.client)
+
+  private def getMajoritySnapshot(peers: Peers): IO[SnapshotInfo] =
+    peers.values
+      .map(peerData => peerData.client)
+      .toList
+      .traverse(client =>
+        client
+          .getNonBlockingBytesKryo[SnapshotInfo]("info", timeout = 15.seconds)
+          .toIO
+      )
+        .map(snapshots => snapshots.groupBy(_.snapshot.hash).maxBy(_._2.size)._2.head)
+
+  private def downloadAndProcessSnapshotsFirstPass(snapshotInfo: SnapshotInfo)(
+    implicit snapshotClient: APIClient,
+    peers: Peers
+  ): IO[Seq[String]] = {
+    val snapshotHashes = getSnapshotHashes(snapshotInfo)
+
+    snapshotHashes
+      .flatMap(getGroupedHashes)
+      .flatMap(processSnapshots)
+      .flatMap(_ => updateMetric("downloadFirstPassComplete", "true"))
+      .flatMap(_ => setNodeState(NodeState.DownloadCompleteAwaitingFinalSync))
+      .flatMap(_ => snapshotHashes)
+  }
 
-    logger.info(s"Downloading from: ${snapshotClient.hostName}:${snapshotClient.apiPort}")
+  private def downloadAndProcessSnapshotsSecondPass(
+    snapshotInfo: SnapshotInfo,
+    hashes: Seq[String]
+  )(implicit snapshotClient: APIClient, peers: Peers): IO[SnapshotInfo] = {
+    val snapshotHashes = getSnapshotHashes(snapshotInfo)
+      .map(_.filterNot(hashes.contains))
+      .flatMap(
+        hashes =>
+          updateMetric("downloadExpectedNumSnapshotsSecondPass", hashes.size.toString)
+            .map(_ => hashes)
+      )
+
+    snapshotHashes
+      .flatMap(getGroupedHashes)
+      .flatMap(processSnapshots)
+      .flatMap(_ => updateMetric("downloadSecondPassComplete", "true"))
+      .map(_ => snapshotInfo)
+  }
 
-    val snapshotInfo =
-      snapshotClient.getBlockingBytesKryo[SnapshotInfo]("info", timeout = 300.seconds)
+  private def finishDownload(snapshot: SnapshotInfo): IO[Unit] = {
+    setSnapshot(snapshot)
```

Comment:
maybe write these long chained flatMap's in for comprehension style?

We could add this plugin if you run into issues: https://github.com/oleg-py/better-monadic-for

Reply:
yup, I can